src/spi.rs

```diff
@@ -1298,9 +1638,124 @@ fn polling_transmit(
         length: (transaction_length * 8) as _,
         rxlength: (rx_length * 8) as _,
         ..Default::default()
-    };
+    }
+}
+
+fn spi_transmit(
+    handle: spi_device_handle_t,
+    transaction: &mut spi_transaction_t,
+    polling: bool,
+    _keep_cs_active: bool,
+) -> Result<(), EspError> {
+    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
+    // The CS pin should be kept active through transactions.
+    #[cfg(not(esp_idf_version = "4.3"))]
+    if _keep_cs_active {
+        transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
+    }
+
+    if polling {
+        esp!(unsafe { spi_device_polling_transmit(handle, transaction as *mut _) })
+    } else {
+        esp!(unsafe { spi_device_transmit(handle, transaction as *mut _) })
+    }
+}
+
+async fn spi_transmit_async(
+    handle: spi_device_handle_t,
+    transaction: &mut spi_transaction_t,
+    _keep_cs_active: bool,
+) -> Result<(), EspError> {
+    // This unfortunately means that this implementation is incorrect for esp-idf < 4.4.
+    // The CS pin should be kept active through transactions.
+    #[cfg(not(esp_idf_version = "4.3"))]
+    if _keep_cs_active {
+        transaction.flags |= SPI_TRANS_CS_KEEP_ACTIVE
+    }
+
+    let notification = Notification::new();
 
-    esp!(unsafe { spi_device_polling_transmit(handle, &mut transaction as *mut _) })
+    transaction.user = ptr::addr_of!(notification) as *mut _;
+
+    loop {
+        match esp!(unsafe {
+            spi_device_queue_trans(handle, transaction as *mut _, delay::NON_BLOCK)
+        }) {
+            Ok(_) => break,
+            Err(e) if e.code() != ESP_ERR_TIMEOUT => return Err(e),
+            _ => notification.wait().await,
+        }
+    }
+
+    Completion::new(
+        async {
+            loop {
+                match esp!(unsafe {
+                    spi_device_get_trans_result(handle, core::ptr::null_mut(), delay::NON_BLOCK)
+                }) {
+                    Ok(_) => break Ok(()),
+                    Err(e) if e.code() != ESP_ERR_TIMEOUT => break Err(e),
+                    _ => notification.wait().await,
+                }
+            }
```

Comment:
You should only call `wait()` once, otherwise this code will run into deadlock. 

```suggestion
            notification.wait().await;
            esp!(unsafe {
                spi_device_get_trans_result(handle, core::ptr::null_mut(), delay::NON_BLOCK)
            })?;
```

Reply:
Ditto. Your code would also work even though I find it riskier. But I don't agree that the existing code can run into a deadlock.

Reply:
In fact if we switch it to look like your code, we should never, ever get an `ESP_ERR_TIMEOUT` or else the code would be wrong. I think this would even need to be asserted. But then I'm not sure we should change it in the first place.

Reply:
Only one transaction has been queued, which means `notification.notify()` will only be called once. The first `notification.wait()` will receive the only `notify()`, any further `notification.wait()`s will wait forever since `notify()` won't be called again.